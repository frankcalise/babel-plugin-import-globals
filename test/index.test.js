import {transform} from '@babel/core';
import transformRuntime from '@babel/plugin-transform-runtime';
import syntaxClassProperties from '@babel/plugin-syntax-class-properties';
import syntaxJsx from '@babel/plugin-syntax-jsx';
import transformJsx from '@babel/plugin-transform-react-jsx';
import plugin from '../src';

test('Invalid options throws helpful error', () => {
  // counter example
  const opts = {
    babelrc: false,
    plugins: [
      [
        plugin,
        {
          Promise: {name: 'promise'},
        },
      ],
    ],
  };
  const input = `Promise.resolve(foo).then(console.log);`;
  expect(
    () => transform(input, opts)
  ).toThrow(/Invalid config options for babel-plugin-import-globals/);
});

test('Imports Promise', () => {
  const opts = {
    babelrc: false,
    plugins: [
      [
        plugin,
        {
          Promise: 'promise',
        },
      ],
    ],
  };
  const input = `Promise.resolve(foo).then(console.log);`;
  const {code} = transform(input, opts);
  expect(code).toMatchSnapshot();
});
test('Still imports Promise when we use transform runtime first', () => {
  const opts = {
    babelrc: false,
    plugins: [
      transformRuntime,
      [
        plugin,
        {
          Promise: {moduleName: 'promise', exportName: 'default'},
        },
      ],
    ],
  };
  const input = `Promise.resolve(foo).then(console.log);`;
  const {code} = transform(input, opts);
  expect(code).toMatchSnapshot();
});
test('Still imports Promise when we use transform runtime after', () => {
  const opts = {
    babelrc: false,
    plugins: [
      [
        plugin,
        {
          Promise: {moduleName: 'promise', exportName: 'default'},
        },
      ],
      transformRuntime,
    ],
  };
  const input = `Promise.resolve(foo).then(console.log);`;
  const {code} = transform(input, opts);
  expect(code).toMatchSnapshot();
});
test('Imports React', () => {
  const opts = {
    babelrc: false,
    plugins: [
      syntaxClassProperties,
      syntaxJsx,
      [
        plugin,
        {
          React: 'react',
          Component: {moduleName: 'react', exportName: 'Component'},
          PropTypes: {moduleName: 'react', exportName: 'PropTypes'},
        },
      ],
    ],
  };
  const input = `
    class MyElement extends Component {
      static propTypes = {
        name: PropTypes.string.isRequired,
      };
      render() {
        return <div>{this.props.name}</div>
      }
    }
  `;
  const {code} = transform(input, opts);
  expect(code).toMatchSnapshot();
});

test('Imports React + JSX', () => {
  const opts = {
    babelrc: false,
    plugins: [
      syntaxClassProperties,
      transformJsx,
      [
        plugin,
        {
          React: 'react',
          Component: {moduleName: 'react', exportName: 'Component'},
          PropTypes: {moduleName: 'react', exportName: 'PropTypes'},
        },
      ],
    ],
  };
  const input = `
    class MyElement extends Component {
      static propTypes = {
        name: PropTypes.string.isRequired,
      };
      render() {
        return <div>{this.props.name}</div>
      }
    }
  `;
  const {code} = transform(input, opts);
  expect(code).toMatchSnapshot();
});


test('Does not double import React + JSX', () => {
  const opts = {
    babelrc: false,
    plugins: [
      syntaxClassProperties,
      transformJsx,
      [
        plugin,
        {
          React: 'react',
          Component: {moduleName: 'react', exportName: 'Component'},
          PropTypes: {moduleName: 'react', exportName: 'PropTypes'},
        },
      ],
    ],
  };
  const input = `
    import React, {Component, PropTypes} from 'react';

    class MyElement extends Component {
      static propTypes = {
        name: PropTypes.string.isRequired,
      };
      render() {
        return <div>{this.props.name}</div>
      }
    }
  `;
  const {code} = transform(input, opts);
  expect(code).toMatchSnapshot();
});

test('JSX Elements', () => {
  const opts = {
    babelrc: false,
    plugins: [
      syntaxClassProperties,
      transformJsx,
      [
        plugin,
        {
          React: 'react',
          Component: {moduleName: 'react', exportName: 'Component'},
          PropTypes: {moduleName: 'react', exportName: 'PropTypes'},
          Match: {moduleName: 'react-router', exportName: 'Match'},
        },
      ],
    ],
  };
  const input = `
    const element = (
      <div>
        <ul>
          <li><Link to="/">Home</Link></li>
          <li><Link to="/about">About</Link></li>
          <li><Link to="/topics">Topics</Link></li>
        </ul>

        <hr/>

        <Match exactly pattern="/" component={Home} />
        <Match pattern="/about" component={About} />
        <Match pattern="/topics" component={Topics} />
      </div>
    );
  `;
  const {code} = transform(input, opts);
  expect(code).toMatchSnapshot();
});

test('Handles naming collisions', () => {
  const opts = {
    babelrc: false,
    plugins: [
      [
        plugin,
        {
          Promise: 'promise',
        },
      ],
    ],
  };
  const input = `
    var _Promise;
    Promise.resolve(foo).then(console.log);
    {
      let _Promise2;
      Promise.resolve(foo).then(console.log);
    }
    function foo() {
      var _Promise3;
      Promise.resolve(foo).then(console.log);
    }
  `;
  const {code} = transform(input, opts);
  expect(code).toMatchSnapshot();
});
