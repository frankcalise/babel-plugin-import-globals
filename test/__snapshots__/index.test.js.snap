// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Does not double import React + JSX 1`] = `
"import React, { Component, PropTypes } from 'react';

class MyElement extends Component {
  static propTypes = {
    name: PropTypes.string.isRequired
  };

  render() {
    return React.createElement(\\"div\\", null, this.props.name);
  }

}"
`;

exports[`Handles naming collisions 1`] = `
"import _Promise4 from \\"promise\\";

var _Promise;

_Promise4.resolve(foo).then(console.log);

{
  let _Promise2;

  _Promise4.resolve(foo).then(console.log);
}

function foo() {
  var _Promise3;

  _Promise4.resolve(foo).then(console.log);
}"
`;

exports[`Imports Promise 1`] = `
"import _Promise from \\"promise\\";

_Promise.resolve(foo).then(console.log);"
`;

exports[`Imports React + JSX 1`] = `
"import { Component as _Component } from \\"react\\";
import _React from \\"react\\";
import { PropTypes as _PropTypes } from \\"react\\";

class MyElement extends _Component {
  static propTypes = {
    name: _PropTypes.string.isRequired
  };

  render() {
    return _React.createElement(\\"div\\", null, this.props.name);
  }

}"
`;

exports[`Imports React 1`] = `
"import { Component as _Component } from \\"react\\";
import { PropTypes as _PropTypes } from \\"react\\";

class MyElement extends _Component {
  static propTypes = {
    name: _PropTypes.string.isRequired
  };

  render() {
    return <div>{this.props.name}</div>;
  }

}"
`;

exports[`JSX Elements 1`] = `
"import { Match as _Match } from \\"react-router\\";
import _React from \\"react\\";

const element = _React.createElement(\\"div\\", null, _React.createElement(\\"ul\\", null, _React.createElement(\\"li\\", null, _React.createElement(Link, {
  to: \\"/\\"
}, \\"Home\\")), _React.createElement(\\"li\\", null, _React.createElement(Link, {
  to: \\"/about\\"
}, \\"About\\")), _React.createElement(\\"li\\", null, _React.createElement(Link, {
  to: \\"/topics\\"
}, \\"Topics\\"))), _React.createElement(\\"hr\\", null), _React.createElement(_Match, {
  exactly: true,
  pattern: \\"/\\",
  component: Home
}), _React.createElement(_Match, {
  pattern: \\"/about\\",
  component: About
}), _React.createElement(_Match, {
  pattern: \\"/topics\\",
  component: Topics
}));"
`;

exports[`Still imports Promise when we use transform runtime after 1`] = `
"import _Promise from \\"promise\\";

_Promise.resolve(foo).then(console.log);"
`;

exports[`Still imports Promise when we use transform runtime first 1`] = `
"import _Promise from \\"promise\\";

_Promise.resolve(foo).then(console.log);"
`;
